File: 0354-russian-doll-envelopes/0354-russian-doll-envelopes.cpp
class Solution {
public:
    static bool cmp(vector<int>&A, vector<int>&B) {
        return A[0] == B[0] ? A[1] > B[1] : A[0] < B[0];
    }
    int maxEnvelopes(vector<vector<int>>& env) {
        sort(env.begin(), env.end(), cmp);
        int n = env.size();
        vector<int>ans;
        ans.push_back(env[0][1]);
        for(int i=1; i<n; i++) {
            if(ans.back() < env[i][1]) {
                ans.push_back(env[i][1]);
            }
            else {
                int ind = lower_bound(ans.begin(), ans.end(), env[i][1]) - ans.begin();
                ans[ind] = env[i][1];
            }
        }
        return ans.size();
    }
};

File: 0354-russian-doll-envelopes/0354-russian-doll-envelopes.java
class Solution {
    public int maxEnvelopes(int[][] env) {
        Arrays.sort(env, (a, b) -> {
            if(a[0] == b[0]) return b[1] - a[1];
            else return a[0] - b[0];
        });
        int n = env.length;
        List<Integer> ans = new ArrayList<>();
        ans.add(env[0][1]);
        for(int i=1; i<n; i++) {
            if(ans.get(ans.size()-1) < env[i][1]) {
                ans.add(env[i][1]);
            }
            else {
                int ind = Collections.binarySearch(ans, env[i][1]);
                if(ind < 0) ind = -(ind + 1);
                ans.set(ind, env[i][1]);
            }
        }
        return (int) ans.size();
    }
}